linsolve/linsolve.py

```diff
@@ -323,78 +340,125 @@ def get_weighted_data(self):
+        there are a couple corner cases where lstsq is unstable but pinv works 
+        for the same rcond. It seems particularly the case for single precision matrices.'''
+        xhat = np.linalg.lstsq(A, y[...,0], rcond=rcond)[0]
+        xhat.shape += (1,)
+        return xhat
+
+    def _gen_AtAiAt(self, A, rcond):
+        '''Helper function for forming (At A)^-1 At.  Uses pinv to invert.'''
+        At = A.T.conj()
+        AtA = At.dot(A)
+        # finding inverse is about 3x slower than solve
+        try: AtAi = np.linalg.pinv(AtA, rcond=rcond)
+        except(np.linalg.LinAlgError): 
+            AtAi = np.linalg.inv(AtA)
+        AtAiAt = AtAi.dot(A.T.conj()) 
+        return AtAiAt
+
+    def _invert_pinv(self, A, y, rcond):
+        '''Use np.linalg.pinv to invert AtA matrix.  Tends to be about ~3x slower than solve.'''
+        AtAiAt = self._gen_AtAiAt(A, rcond)
+        return np.dot(AtAiAt,y)
+
+    def _invert_solve(self, A, y, rcond):
+        '''Use np.linalg.solve to solve a system of equations.  Requires a fully constrained
+        system of equations (i.e. doesn't deal with singular matrices).  Can by ~1.5x faster that lstsq
+        for this case.'''
+        At = A.T.conj()
+        AtA = At.dot(A)
+        Aty = At.dot(y)
+        return np.linalg.solve(AtA, Aty) # is supposed to error if singular, but doesn't seem to.
+
+    def _invert_default(self, A, y, rcond):
+        '''Use the lsqr inverter, but if an LinAlgError is encountered, try using pinv.'''
+        try:
+            xhat = self._invert_lsqr(A, y, rcond)
+        except(np.linalg.LinAlgError):
+            xhat = self._invert_pinv(A, y, rcond)
+        return xhat
+
+    def solve(self, rcond=1e-10, mode='default', verbose=False):
         """Compute x' = (At A)^-1 At * y, returning x' as dict of prms:values.
 
         Args:
             rcond: cutoff ratio for singular values useed in numpy.linalg.lstsq, numpy.linalg.pinv,
                 or (if sparse) as atol and btol in scipy.sparse.linalg.lsqr
+            mode: 'default', 'lsqr', 'pinv', or 'solve', selects what inverter to use.  'default'
+                is best for most cases, but 'solve' may be faster for fully-constrained systems,
+                and 'pinv' can sometimes be more stable (but slower) than 'lsqr'.
+            verbose: print information about iterations
 
         Returns:
             sol: a dictionary of solutions with variables as keys
         """
+        assert(mode in ['default','lsqr','pinv','solve'])
         y = self.get_weighted_data()
         Ashape = self._A_shape()
         x = np.empty((Ashape[1],y.shape[-1]), dtype=self.dtype)
         if self.sparse:
+            if mode == 'default': _invert = self._invert_default_sparse
+            elif mode == 'lsqr': _invert = self._invert_lsqr_sparse
+            elif mode == 'pinv': _invert = self._invert_pinv_sparse
+            elif mode == 'solve': _invert = self._invert_solve_sparse
```

Comment:
you probably want an else: raise error. Otherwise it'll just say `_invert` is not defined, which is a rather opaque error message.

Reply:
the "assert" at the top of the code block ensures it's one of these 4 options.

Reply:
Ah, I missed that. OK.